Sources/Networking/APIRequest.swift

```diff
@@ -0,0 +1,115 @@
+//
+//  APIRequest.swift
+//  DuckDuckGo
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import os.log
+
+public typealias APIResponse = (data: Data?, response: HTTPURLResponse)
+public typealias APIRequestCompletion = (APIResponse?, APIRequest.Error?) -> Void
+
+public struct APIRequest {
+    
+    private let request: URLRequest
+    private let requirements: APIResponseRequirements
+    private let urlSession: URLSession
+    private let log: OSLog
+    
+    public init<QueryParams: Collection>(configuration: APIRequest.Configuration<QueryParams>,
+                                         requirements: APIResponseRequirements = [],
+                                         urlSession: URLSession = .shared,
+                                         log: OSLog = .disabled) {
+        self.request = configuration.request
+        self.requirements = requirements
+        self.urlSession = urlSession
+        self.log = log
+        
+        assertUserAgentIsPresent()
+    }
+    
+    private func assertUserAgentIsPresent() {
+        guard request.allHTTPHeaderFields?[HTTPHeaderField.userAgent] != nil else {
+            assertionFailure("A user agent must be included in the request's HTTP header fields.")
+            return
+        }
+    }
+
+    /// This method is deprecated. Please use the 'fetch()' async method instead.
```

Comment:
If it’s only used on iOS, perhaps worth guarding this function with `#if os(iOS)`?
Or maybe add the deprecation notice to the os_log too?